---
title: "break features, CAST"
author: "Yue"
date: "8/10/2018"
output: html_document
---

```{r}
library(tidyverse)
library(sciliantifig)
library(caret)
library(pROC)
library(BiocParallel)
```

# break points and controls

```{r}
hb_cast_df <- read_tsv(slfile("haploid.no0.hb.filtered.cast.tab"))
hb_cast_gc <- read_tsv("~/Desktop/sci-lianti-files/features/cast/haploid_crossover.extended10b.gc.txt")

ctrl_cast_df <- read_tsv("~/Desktop/sci-lianti-files/features/cast/sample.srt.bed", col_names=c("CHROM", "START", "END", "TYPE"))
ctrl_cast_gc <- read_tsv("~/Desktop/sci-lianti-files/features/cast/sample.srt.gc.txt")

hb_ctrl_cast_df <- filter(ctrl_cast_df, TYPE == "hp") %>% dplyr::select(-TYPE)
hb_ctrl_cast_gc <- ctrl_cast_gc[which(ctrl_cast_df$TYPE == "hp"), ]

m2_cast_df <- read_tsv(slfile("m2diploid.concensus.mb.cast.tab"))
m2_cast_gc <- read_tsv("~/Desktop/sci-lianti-files/features/cast/m2.cast.final.extended10b.gc.txt")

m2_ctrl_cast_df <- filter(ctrl_cast_df, TYPE == "m2") %>% dplyr::select(-TYPE)
m2_ctrl_cast_gc <- ctrl_cast_gc[which(ctrl_cast_df$TYPE == "m2"), ]

window_df <- read_tsv("~/Desktop/sci-lianti-files/mm10.window.noXY.bed", col_names=c("CHROM", "START", "END"))
window_gc <- read_tsv("~/Desktop/sci-lianti-files/mm10.window.noXY.gc.txt")

mm10_fai_df <- read_tsv("~/Desktop/sci-lianti-files/mm10.fa.fai", col_names=c("CHROM", "CHROM_SIZE", "BI", "BASES_PER_LINE", "BYTES_PER_LINE"))
```

# features

```{r}
cast_features <- list(
  SPO11 = read_tsv("~/Desktop/sci-lianti-files/features/B6.Spo11.withHits.txt") %>%
    dplyr::select(chr, Start, End, Hits) %>%
    dplyr::rename(CHROM=chr, START=Start, END=End, SPO11=Hits),
  SSDS_F1 = distinct(read_tsv("~/Desktop/sci-lianti-files/features/cast/ssds_b6_cast.tab", col_names=c("CHROM", "START", "END", "SSDS_F1"))),
  SSDS_B6 = distinct(read_tsv("~/Desktop/sci-lianti-files/features/cast/ssds_b6.tab", col_names=c("CHROM", "START", "END", "SSDS_B6"))),
  SSDS_CAST = distinct(read_tsv("~/Desktop/sci-lianti-files/features/cast/ssds_cast.tab", col_names=c("CHROM", "START", "END", "SSDS_CAST"))),
  
  CNV_CAST_GAIN = read_tsv("~/Desktop/sci-lianti-files/features/cast/CNV_CAST_gain.bed", skip=1, col_names=c("CHROM", "START", "END", "CNV_CAST_GAIN")),
  
  CNV_CAST_LOSS = read_tsv("~/Desktop/sci-lianti-files/features/cast/CNV_CAST_loss.bed", skip=1, col_names=c("CHROM", "START", "END", "CNV_CAST_LOSS")) %>%
    mutate(CNV_CAST_LOSS=-CNV_CAST_LOSS),
  
  THREE_PRIME_UTR = read_tsv("~/Desktop/sci-lianti-files/features/mm10.gff3.UTR3.bed", guess_max=1e6),
  
  FIVE_PRIME_UTR = read_tsv("~/Desktop/sci-lianti-files/features/mm10.gff3.UTR5.bed", guess_max=1e6),
  
  GENE = read_tsv("~/Desktop/sci-lianti-files/features/mm10.gff3.gene.bed", guess_max=1e6),
  
  LNC_RNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.gff3.lnc_RNA.bed", guess_max=1e6),
  
  NCRNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.gff3.ncRNA.bed", guess_max=1e6),
  
  PSEUDOGENIC_TRANSCRIPT = read_tsv("~/Desktop/sci-lianti-files/features/mm10.gff3.pseudogenic_transcript.bed", guess_max=1e6),
  
  RRNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.gff3.rRNA.bed", guess_max=1e6),
  
  SNRNA_SNORNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.gff3.snRNA_snoRNA.bed", guess_max=1e6),
  
  RMSK_DNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.DNA.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_DNA=1),
  
  RMSK_LINE = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.LINE.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_LINE=1),
  
  RMSK_LTR = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.LTR.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_LTR=1),
  
  RMSK_LC = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.Low_complexity.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_LC=1),
  
  RMSK_RC = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.RC.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_RC=1),
  
  RMSK_RNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.RNA.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_RNA=1),
  
  RMSK_SINE = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.SINE.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_SINE=1),
  
  RMSK_SATE = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.Satellite.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_SATE=1),
  
  RMSK_SIMPLE_REPEAT = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.Simple_repeat.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_SIMPLE_REPEAT=1),
  
  RMSK_RRNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.rRNA.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_RRNA=1),
  
  RMSK_SCRNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.scRNA.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_SCRNA=1),
  
  RMSK_SNRNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.snRNA.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_SNRNA=1),
  
  RMSK_SRPRNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.srpRNA.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_SRPRNA=1),
  
  RMSK_TRNA = read_tsv("~/Desktop/sci-lianti-files/features/mm10.rmsk.tRNA.bed", col_names=FALSE) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(RMSK_TRNA=1),
  
  SEG_DUP = read_tsv("~/Desktop/sci-lianti-files/features/mm10.segdup.bed", skip=1, col_names=F) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3) %>%
    dplyr::select(CHROM, START, END) %>%
    mutate(SEG_DUP=1),
  
  SEQ_DIV_CAST = read_tsv("~/Desktop/sci-lianti-files/features/cast/CAST.windowed.pi.bedGraph", col_names=c("CHROM", "START", "END", "SEQ_DIV_CAST")),
  
  CTCF = read_tsv("~/Desktop/sci-lianti-files/features/mm10_testis_ctcf.broadPeak.bed", col_names=FALSE) %>%
    dplyr::select(X1, X2, X3, X7) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3, CTCF=X7),
  
  RNA_POL2 = read_tsv("~/Desktop/sci-lianti-files/features/mm10_testis_RNApol2.broadPeak.bed", col_names=FALSE) %>%
    dplyr::select(X1, X2, X3, X7) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3, RNA_POL2=X7),
  
  H3K27AC = read_tsv("~/Desktop/sci-lianti-files/features/mm10_testis_h3k27ac.broadPeak.bed", col_names=FALSE) %>%
    dplyr::select(X1, X2, X3, X7) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3, H3K27AC=X7),
  
  H3K36ME3 = read_tsv("~/Desktop/sci-lianti-files/features/mm10_testis_h3k36me3.broadPeak.bed", col_names=FALSE) %>%
    dplyr::select(X1, X2, X3, X7) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3, H3K36ME3=X7),
  
  H3K4ME1 = read_tsv("~/Desktop/sci-lianti-files/features/mm10_testis_h3k4me1.broadPeak.bed", col_names=FALSE) %>%
    dplyr::select(X1, X2, X3, X7) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3, H3K4ME1=X7),
  
  H3K4ME3 = read_tsv("~/Desktop/sci-lianti-files/features/mm10_testis_h3k4me3.broadPeak.bed", col_names=FALSE) %>%
    dplyr::select(X1, X2, X3, X7) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3, H3K4ME3=X7),
  
  LONG_RNA_SEQ = read_tsv("~/Desktop/sci-lianti-files/features/mm10_testis_longRNAseq.broadPeak.bed", col_names=FALSE) %>%
    dplyr::select(X1, X2, X3, X7) %>%
    dplyr::rename(CHROM=X1, START=X2, END=X3, LONG_RNA_SEQ=X7),
  
  DMRT6 = read_tsv("~/Desktop/sci-lianti-files/features/DMRT6/DMRT6.mm10.bed", col_names=c("CHROM", "START", "END", "DMRT6")),
  
  H3K27ME3 = read_tsv("~/Desktop/sci-lianti-files/features/h3k27me3/h3k27me3.mm10.merged.bed", col_names=c("CHROM", "START", "END")) %>%
    mutate(H3K27ME3=1),
  
  PS_ATAC= read_tsv("~/Desktop/sci-lianti-files/features/PS_atac_H3K27ac/ps.atac.bed", col_names=c("CHROM", "START", "END", "PS_ATAC")),
  
  PS_H3K27AC = read_tsv("~/Desktop/sci-lianti-files/features/PS_atac_H3K27ac/ps.h3k27ac.bed", col_names=c("CHROM", "START", "END", "PS_H3K27AC")),
  
  PS_COHESIN = read_tsv("~/Desktop/sci-lianti-files/features/PS_cohesin/PS_cohesin_0.001.mm10.bed", col_names=c("CHROM", "START", "END", "PS_COHESIN")),
  
  PS_SMC1B_LAP = read_tsv("~/Desktop/sci-lianti-files/features/PS_cohesin/PS_SMC1B-LAP_0.1.mm10.bed", col_names=c("CHROM", "START", "END", "PS_SMC1B_LAP")),
  
  PS_H2AUB = read_tsv("~/Desktop/sci-lianti-files/features/PS_H2Aub/PS_H2Aub.mm10.bed", col_names=c("CHROM", "START", "END", "PS_H2AUB")),
  
  H4K5AC = read_tsv("~/Desktop/sci-lianti-files/features/PS_h4mods/h4k5ac.mm10.bed", col_names=c("CHROM", "START", "END", "H4K5AC")),
  
  H4K5BU = read_tsv("~/Desktop/sci-lianti-files/features/PS_h4mods/h4k5bu.mm10.bed", col_names=c("CHROM", "START", "END", "H4K5BU")),
  
  H4K8AC = read_tsv("~/Desktop/sci-lianti-files/features/PS_h4mods/h4k8ac.mm10.bed", col_names=c("CHROM", "START", "END", "H4K8AC")),
  
  H4K8BU = read_tsv("~/Desktop/sci-lianti-files/features/PS_h4mods/h4k8bu.mm10.bed", col_names=c("CHROM", "START", "END", "H4K8BU")),
  
  CTCFL = read_tsv("~/Desktop/sci-lianti-files/features/CTCFL/CTCFL.mm10.bed", col_names=c("CHROM", "START", "END")) %>%
    mutate(CTCFL=1),
  
  MESC_REPLI_SEQ = read_tsv("~/Desktop/sci-lianti-files/features/mESC_repli_seq/mESC_repli_seq.mm10.bed", col_names=c("CHROM", "START", "END", "MESC_REPLI_SEQ")),
  
  CTCF_B6 = read_tsv("~/Desktop/sci-lianti-files/features/end_seq/CTCF_b6_aB.mm10.bed", col_names=c("CHROM", "START", "END", "CTCF_B6")),
  
  # CTCF_SPRET = read_tsv("~/Desktop/sci-lianti-files/features/end_seq/CTCF_spret_aB.mm10.bed", col_names=c("CHROM", "START", "END", "CTCF_SPRET")),
  
  END_SEQ_B6_ETO = read_tsv("~/Desktop/sci-lianti-files/features/end_seq/end_seq_b6_ETO.mm10.bed", col_names=c("CHROM", "START", "END", "END_SEQ_B6_ETO")),
  
  # END_SEQ_SPRET_ETO = read_tsv("~/Desktop/sci-lianti-files/features/end_seq/end_seq_spret_ETO.mm10.bed", col_names=c("CHROM", "START", "END", "END_SEQ_SPRET_ETO")),
  
  RAD21_B6 = read_tsv("~/Desktop/sci-lianti-files/features/end_seq/RAD21_b6_aB.mm10.bed", col_names=c("CHROM", "START", "END", "RAD21_B6")),
  
  # RAD21_SPRET = read_tsv("~/Desktop/sci-lianti-files/features/end_seq/RAD21_spret_aB.mm10.bed", col_names=c("CHROM", "START", "END", "RAD21_SPRET")),
  
  TOP2A_MEF = read_tsv("~/Desktop/sci-lianti-files/features/end_seq/TOP2A_MEF.mm10.bed", col_names=c("CHROM", "START", "END", "TOP2A_MEF")),
  
  TOP2B_MEF = read_tsv("~/Desktop/sci-lianti-files/features/end_seq/TOP2B_MEF.mm10.bed", col_names=c("CHROM", "START", "END", "TOP2B_MEF")),
  
  PATSKI_ALLELIC_ATAC_B6 = read_tsv("~/Desktop/sci-lianti-files/features/patski/patski_allelicATAC.mm10.b6.bed", col_names=c("CHROM", "START", "END", "PATSKI_ALLELIC_ATAC_B6")),
  
  # PATSKI_ALLELIC_ATAC_SPRET = read_tsv("~/Desktop/sci-lianti-files/features/patski/patski_allelicATAC.mm10.spret.bed", col_names=c("CHROM", "START", "END", "PATSKI_ALLELIC_ATAC_SPRET")),
  
  PATSKI_ALLELIC_CTCF_B6 = read_tsv("~/Desktop/sci-lianti-files/features/patski/patski_allelicCTCF.mm10.b6.bed", col_names=c("CHROM", "START", "END", "PATSKI_ALLELIC_CTCF_B6")),
  
  # PATSKI_ALLELIC_CTCF_SPRET = read_tsv("~/Desktop/sci-lianti-files/features/patski/patski_allelicCTCF.mm10.spret.bed", col_names=c("CHROM", "START", "END", "PATSKI_ALLELIC_CTCF_SPRET")),
  
  # PATSKI_CTCF_B6_SUMMITS = read_tsv("~/Desktop/sci-lianti-files/features/patski/Patski_ctcf_b6_summits.mm10.bed", col_names=c("CHROM", "START", "END", "NOUSE", "PATSKI_CTCF_B6_SUMMITS")) %>% dplyr::select(-NOUSE),
  
  # PATSKI_CTCF_SPRET_SUMMITS = read_tsv("~/Desktop/sci-lianti-files/features/patski/Patski_ctcf_spret_summits.mm10.bed", col_names=c("CHROM", "START", "END", "NOUSE", "PATSKI_CTCF_SPRET_SUMMITS")) %>%
    # select(-NOUSE),
  
  PATSKI_POLIIS5P_B6_SUMMITS = read_tsv("~/Desktop/sci-lianti-files/features/patski/Patski_PolIIS5p_b6_summits.mm10.bed", col_names=c("CHROM", "START", "END", "NOUSE","PATSKI_POLIIS5P_B6_SUMMITS")) %>% dplyr::select(-NOUSE)
  
  # PATSKI_POLIIS5P_SPRET_SUMMITS = read_tsv("~/Desktop/sci-lianti-files/features/patski/Patski_PolIIS5p_spret_summits.mm10.bed", col_names=c("CHROM", "START", "END", "NOUSE", "PATSKI_POLIIS5P_SPRET_SUMMITS")) %>%
    # select(-NOUSE)
)

cast_features <- map(cast_features, ~drop_na(.x))
```

```{r}
cast_other_features <- list(
  EVENT = list(
    P7GONIA = read_tsv("~/Desktop/sci-lianti-files/features/cast/p7gonia_event.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "P7GONIA")),
    PS_MNASE_NUC = read_tsv("~/Desktop/sci-lianti-files/features/cast/PS_MNase_nuc_mm10_event.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "PS_MNASE_NUC")),
    PS_5MC_EVENT = read_tsv("~/Desktop/sci-lianti-files/features/cast/PS_5mc_mm10_event.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "PS_5MC")),
    PS_BS_EVENT = read_tsv("~/Desktop/sci-lianti-files/features/cast/PS_bs_event.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "PS_BS")),
    KAC_EVENT = read_tsv("~/Desktop/sci-lianti-files/features/cast/Kac_event.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "KAC"))
    # KCRO_EVENT = read_tsv("~/Desktop/sci-lianti-files/features/cast/Kcro_event.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "KCRO"))
  ),
  CONTROL = list(
    P7GONIA_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/p7gonia_control.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "P7GONIA")),
    PS_MNASE_NUC_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/PS_MNase_nuc_mm10_control.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "PS_MNASE_NUC")),
    PS_5MC_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/PS_5mc_mm10_control.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "PS_5MC")),
    PS_BS_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/PS_bs_control.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "PS_BS")),
    KAC_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/Kac_control.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "KAC"))
    # KCRO_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/Kcro_control.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "KCRO"))
  ),
  WINDOW = list(
    P7GONIA_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/p7gonia_window.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "P7GONIA")),
    PS_MNASE_NUC_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/PS_MNase_nuc_mm10_window.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "PS_MNASE_NUC")),
    PS_5MC_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/PS_5mc_mm10_window.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "PS_5MC")),
    PS_BS_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/PS_bs_window.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "PS_BS")),
    KAC_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/Kac_window.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "KAC")),
    # KCRO_CONTROL = read_tsv("~/Desktop/sci-lianti-files/features/cast/Kcro_window.cast.bed", col_names=c("CHROM", "START", "END", "CELL_ID", "KCRO")),
    event_hp_m2 = read_tsv("~/Desktop/sci-lianti-files/features/cast/event_window.bed", col_names=c("CHROM", "START", "END", "event_hp_m2")),
    event_hp = read_tsv("~/Desktop/sci-lianti-files/features/cast/haploid_window.cast.bed", col_names=c("CHROM", "START", "END", "event_hp")),
    event_m2 = read_tsv("~/Desktop/sci-lianti-files/features/cast/m2_window.cast.bed", col_names=c("CHROM", "START", "END", "event_m2")),
    event_mt = read_tsv("~/Desktop/sci-lianti-files/features/cast/mt_window.cast.bed", col_names=c("CHROM", "START", "END", "event_mt")),
    event_me = read_tsv("~/Desktop/sci-lianti-files/features/cast/me_window.cast.bed", col_names=c("CHROM", "START", "END", "event_me"))
  )
)

cast_other_features <- map(cast_other_features, function(x) {
  map(x, function(y) {
    signal_col <- colnames(y)[min(5, ncol(y))]
    # signal_col <- rlang::sym(colnames(y)[5])
    if (!is.numeric(y[[signal_col]])) {
      y[[signal_col]][y[[signal_col]] == '.'] <- '0'
      y[[signal_col]] <- as.numeric(y[[signal_col]])
    }
    y
  })
})

cast_other_features_processed <- list(
  EVENT_HP = map(cast_other_features$EVENT, function(x) {
    right_join(x, hb_cast_df[c("CELL_ID", "CHROM", "START", "END")])
  }),
  EVENT_M2 = map(cast_other_features$EVENT, function(x) {
    right_join(x, m2_cast_df[c("CELL_ID", "CHROM", "START", "END")])
  }),
  CONTROL_HP = map(cast_other_features$CONTROL, function(x) {
    filter(x, CELL_ID == "hp")
  }),
  CONTROL_M2 = map(cast_other_features$CONTROL, function(x) {
    filter(x, CELL_ID == "m2")
  }),
  WINDOW = cast_other_features$WINDOW
)
```

# per track feature summary

```{r}
add_column_by_overlap_join <- function(old_df, df_to_add) {
  require(data.table)
  remove_cellid_before_return <- FALSE
  if (!"CELL_ID" %in% colnames(old_df)) {
    # for ease of future joining, add a unique CELL_ID to each row so it doesn't contribute to group_by
    old_df <- add_column(old_df, CELL_ID=c(1:nrow(old_df)), .before=1)
    remove_cellid_before_return <- TRUE
  }
  # old_df: CELL_ID CHROM START END some other fields
  assertthat::are_equal(colnames(old_df)[1:4], c("CELL_ID", "CHROM", "START", "END"))
  # df_to_add: CHROM START END signal (where signal is the column to be added by overlap join and average)
  assertthat::are_equal(colnames(df_to_add)[1:3], c("CHROM", "START", "END"))
  signal_col <- rlang::sym(colnames(df_to_add)[4])
  if (mean(df_to_add$START == df_to_add$END) > 0.95) df_to_add$END <- df_to_add$END + 1
  df_to_add <- df_to_add[!is.na(df_to_add$START) & !is.na(df_to_add$END) & df_to_add$START < df_to_add$END, ]
  numeric_signal <- TRUE
  if (mean(df_to_add[[signal_col]] == 1) > 0.95) numeric_signal <- FALSE
  x <- data.table(df_to_add)
  y <- data.table(old_df)
  setkey(y, CHROM, START, END)
  yx <- foverlaps(x, y, by.x=c("CHROM", "START", "END"), by.y=c("CHROM", "START", "END"), type="any")
  yx_summary <- as.tibble(yx) %>%
    drop_na(START, END)
  if (numeric_signal) {
    yx_summary <- yx_summary %>%
      group_by(CELL_ID, CHROM, START, END) %>%
      summarise(!!signal_col:=mean(!!signal_col))
  } else {
    yx_summary <- yx_summary %>%
      group_by(CELL_ID, CHROM, START, END) %>%
      summarise(!!signal_col:=sum(!!signal_col))
  }
  ret <- old_df %>%
    left_join(yx_summary, by=c("CELL_ID", "CHROM", "START", "END"))
  assertthat::are_equal(nrow(ret), nrow(old_df))
  # by default set tracks with no overlaps found to 0
  ret[is.na(ret[[colnames(df_to_add)[4]]]), colnames(df_to_add)[4]] <- 0
  if (remove_cellid_before_return) ret <- ret[colnames(ret) != "CELL_ID"]
  return(ret)
}

add_column_by_join <- function(old_df, df_to_add) {
  # add column by a simple join
  signal_col <- colnames(df_to_add)[min(5, ncol(df_to_add))]
  if ("CELL_ID" %in% colnames(old_df)) {
    # real data with real cell id
    assertthat::assert_that(length(unique(df_to_add$CELL_ID)) > 10)
    ret <- left_join(old_df, df_to_add[c("CELL_ID", "CHROM", "START", "END", signal_col)], by=c("CELL_ID", "CHROM", "START", "END"))
    assertthat::are_equal(nrow(ret), nrow(old_df))
    return(ret)
  }
  # simulated data without cell id
  ret <- left_join(old_df, df_to_add[c("CHROM", "START", "END", signal_col)], by=c("CHROM", "START", "END"))
  assertthat::are_equal(nrow(ret), nrow(old_df))
  return(ret)
}

process_features <- function(break_points_df, gc, features_to_overlap, features_to_join) {
  bp <- break_points_df %>%
    mutate(BP_MID=round((START + END) / 2)) %>%
    left_join(mm10_fai_df[c("CHROM", "CHROM_SIZE")], by="CHROM") %>%
    mutate(QUANTILE_25=(CHROM_SIZE - 3e6) / 4 + 3e6,
           QUANTILE_50=(CHROM_SIZE - 3e6) / 2 + 3e6,
           QUANTILE_75=(CHROM_SIZE - 3e6) / 4 * 3 + 3e6) %>%
    mutate(CENTROMERIC=case_when(BP_MID < 1e7 + 3e6 ~ "Yes", TRUE ~ "No"),
           TELOMERIC=case_when(BP_MID > CHROM_SIZE - 1e7 ~ "Yes", TRUE ~ "No"),
           QUANTILE_0_25_FLAG=case_when(BP_MID < QUANTILE_25 ~ "Yes", TRUE ~ "No"),
           QUANTILE_25_75_FLAG=case_when(BP_MID >= QUANTILE_25 & BP_MID < QUANTILE_75 ~ "Yes", TRUE ~ "No"),
           QUANTILE_75_100_FLAG=case_when(BP_MID >= QUANTILE_75 ~ "Yes", TRUE ~ "No"))
  assertthat::assert_that(mean(gc$`2_usercol` == bp$START) > 0.95)
  gc_col <- grep("pct_gc", colnames(gc), value=TRUE)
  bp$GC <- gc[[gc_col]]
  bp_f <- bp
  for (f in features_to_overlap) {
    bp_f <- add_column_by_overlap_join(bp_f, f)
  }
  for (f in features_to_join) {
    bp_f <- add_column_by_join(bp_f, f)
  }
  bp_f$DISTANCE <- bp_f$END - bp_f$START
  bp_f$KAC <- bp_f$KAC / (bp_f$DISTANCE + 1000)
  # bp_f$KCRO <- bp_f$KCRO / (bp_f$DISTANCE + 1000)
  assertthat::are_equal(nrow(break_points_df), nrow(bp_f))
  return(bp_f)
}
```

# generate per tract data

```{r}
hb_cast_df_with_cor <- process_features(hb_cast_df, hb_cast_gc, cast_features, cast_other_features_processed$EVENT_HP)
m2_cast_df_with_cor <- process_features(m2_cast_df, m2_cast_gc, cast_features, cast_other_features_processed$EVENT_M2)
hb_cast_ctrl_with_cor <- process_features(hb_ctrl_cast_df, hb_ctrl_cast_gc, cast_features, cast_other_features_processed$CONTROL_HP)
m2_cast_ctrl_with_cor <- process_features(m2_ctrl_cast_df, m2_ctrl_cast_gc, cast_features, cast_other_features_processed$CONTROL_M2)

window_cast_with_cor <- process_features(window_df, window_gc, cast_features, cast_other_features_processed$WINDOW)

haploid_cast_upd <- read_tsv(slfile("haploid.upd.cast.tab"))

m2_cast_upd <- read_tsv(slfile("m2.upd.cast.tab"))
# write_tsv(hb_df_with_cor, "~/Desktop/sci-lianti-files/haploid_breakpoints_with_feature.tab")
# write_tsv(m2_df_with_cor, "~/Desktop/sci-lianti-files/m2diploid_breakpoints_with_feature.tab")
```

# per cell feature summary

```{r}
# here, source relevant function from the spret part

hb_cast_cell_df_with_cor_ <- per_cell_summary(hb_cast_df_with_cor, haploid_cast_upd, num_cols, int_cols)
m2_cast_cell_df_with_cor_ <- per_cell_summary(m2_cast_df_with_cor, m2_cast_upd, num_cols, int_cols)

# use SUM_m2_mt as n_mt, use SUM_m2_me as n_me, for haploids, n_mt=0, n_me=19 for all the cells
m2_cast_mt_me <- read_tsv(slfile("cast/all_m2_merge.cast.tab")) %>%
  filter(TYPE != "Others")
m2_cast_mt_me_df <- tibble(CELL_ID=m2_cast_mt_me$CELL_ID, N_MT=m2_cast_mt_me$SUM_m2_mt, N_ME=m2_cast_mt_me$SUM_m2_me)

hb_cast_cell_df_with_cor <- mutate(hb_cast_cell_df_with_cor_, N_MT=0, N_ME=19)
m2_cast_cell_df_with_cor <- left_join(m2_cast_cell_df_with_cor_, m2_cast_mt_me_df, by="CELL_ID")

cast_cell_df_with_cor <- bind_rows(hb_cast_cell_df_with_cor %>% add_column(TYPE="haploid", .after=1),
                                    m2_cast_cell_df_with_cor %>% add_column(TYPE="m2 diploid", .after=1))
```

```{r}
save(hb_cast_df_with_cor,
     m2_cast_df_with_cor,
     hb_cast_ctrl_with_cor,
     m2_cast_ctrl_with_cor,
     window_cast_with_cor,
     cast_cell_df_with_cor,
     file="~/Desktop/sciliantifig/inst/extdata/cast_features.rda")
```


# models

```{r}
md_cast_data <- bind_rows(hb_cast_df_with_cor %>% add_column(RESPONSE="Y", TYPE="hp", .before=1),
                          m2_cast_df_with_cor %>% add_column(RESPONSE="Y", TYPE="m2", .before=1),
                          hb_cast_ctrl_with_cor %>% add_column(RESPONSE="N", TYPE="hp", .before=1),
                          m2_cast_ctrl_with_cor %>% add_column(RESPONSE="N", TYPE="m2", .before=1)) %>%
  select(-c(CELL_ID, CHROM, START, END, SCORE, STRAND, DISTANCE, BP_MID, CHROM_SIZE, QUANTILE_25, QUANTILE_50, QUANTILE_75, NOTE, SEGREGATION)) # %>%
  # drop_na()
saveRDS(md_cast_data, "~/Desktop/sci-lianti-files/180811_md_cast_data.rds")

# hb_data <- filter(md_data, TYPE == "HB") %>%
#   select(-TYPE)
# md_data <- md09_data
md_data <- md_cast_data

ncv_list <- bplapply(1:2, function(x) {
  train_n <- round(nrow(md_data) * 0.9)
  train_idx <- sample(1:nrow(md_data), size=train_n, replace=FALSE)
  train_df <- md_data[train_idx, ]
  train_df$RESPONSE <- factor(train_df$RESPONSE)
  test_df <- md_data[-train_idx, ]
  train_control <- trainControl(method = "cv", number = 5, returnResamp = "all",
                                classProbs = TRUE, 
                                summaryFunction = twoClassSummary)
  model_weights <- ifelse(train_df$RESPONSE == "Y",
                          (1/table(train_df$RESPONSE)["Y"]) * 0.5,
                          (1/table(train_df$RESPONSE)["N"]) * 0.5)
  md <- train(RESPONSE ~ ., data = train_df, 
              method = "rf", 
              trControl = train_control,
              weights = model_weights,
              metric = "ROC",
              na.action = "na.omit")
  pred <- predict(md, newdata=test_df, type = "prob", na.action="na.omit")
  pred_df <- tibble(truth=test_df[complete.cases(test_df), ]$RESPONSE, pred=pred$Y)
  return(pred_df)
})

# saveRDS(pred_df, "180812_md_cast_pred.rds")
# save.image("180812_md_cast_pred_rf.RData")

pred_df <- readRDS("~/Desktop/sci-lianti-files/180812_md_cast_pred.rds")

# use only MIP > 0.5 features

use_cast_cols <- c("RESPONSE","CENTROMERIC","TELOMERIC","QUANTILE_0_25_FLAG","QUANTILE_25_75_FLAG","GC","SSDS_F1","SSDS_B6","SSDS_CAST","CNV_CAST_GAIN","THREE_PRIME_UTR","GENE","PSEUDOGENIC_TRANSCRIPT","RMSK_DNA","RMSK_LINE","RMSK_LC","RMSK_SIMPLE_REPEAT","SEQ_DIV_CAST","H3K4ME1","H3K4ME3","DMRT6","H3K27ME3","MESC_REPLI_SEQ","TOP2A_MEF","PATSKI_ALLELIC_CTCF_B6","P7GONIA")

sub_train_df <- train_df[use_cast_cols]
sub_test_df <- test_df[use_cast_cols]

sub_md <- train(RESPONSE ~ ., data = sub_train_df, 
                method = "rf", 
                trControl = train_control,
                weights = model_weights,
                metric = "ROC",
                na.action = "na.omit")
sub_pred <- predict(sub_md, newdata=sub_test_df, type = "prob", na.action="na.omit")
sub_pred_df <- tibble(truth=sub_test_df[complete.cases(sub_test_df), ]$RESPONSE, pred=sub_pred$Y)

# saveRDS(sub_pred_df, "180813_md_cast_sub_pred.rds")
# sub_pred_df <- readRDS("~/Desktop/sci-lianti-files/180813_md_cast_sub_pred.rds")

ncv_df <- bind_rows(ncv_list)
roc_obj <- roc(ncv_df$truth, ncv_df$pred)
plot(roc_obj)
auc(roc_obj)

roc_obj <- roc(sub_pred_df$truth, sub_pred_df$pred)
plot(roc_obj)
auc(roc_obj)

p <- ggplot(tibble(Sensitivity=roc_obj$sensitivities, Specificity=roc_obj$specificities), aes(Specificity, Sensitivity)) +
  geom_segment(x=0, y=1, xend=-1, yend=0, color="grey50", size=1) +
  geom_line(color="orangered", size=1) +
  xlim(1, 0) +
  ylim(0, 1) +
  theme_classic()

ggsave("~/Desktop/sci-lianti-files/cast_rf_roc_sub.pdf", p, "pdf", width=2.6, height=2.5)
```

```{r}
as.numeric.factor <- function(x) {as.numeric(levels(x))[x]}

cast_window_data <- window_cast_with_cor %>%
  dplyr::select(-c(CHROM, START, END, DISTANCE, BP_MID, CHROM_SIZE, QUANTILE_25, QUANTILE_50, QUANTILE_75)) %>%
  mutate_if(sapply(., is.character), function(x) {ifelse(x == "Yes", 1, 0)})
cast_window_data <- cast_window_data[!grepl("event_", colnames(cast_window_data)) | colnames(cast_window_data) == "event_hp_m2"]

cast_window_lm <- lm(log1p(event_hp_m2) ~ ., data=cast_window_data)

sink("~/Desktop/sci-lianti-files/cast_window_hp_m2_lm_output.txt")
print(summary(cast_window_lm))
sink()

window_hp_lm <- lm(event_hp ~ ., data=window_data[!grepl("event_", colnames(window_data)) | colnames(window_data) == "event_hp"])
window_m2_lm <- lm(event_m2 ~ ., data=window_data[!grepl("event_", colnames(window_data)) | colnames(window_data) == "event_m2"])
# window_mt_lm <- lm(event_mt ~ ., data=window_data[!grepl("event_", colnames(window_data)) | colnames(window_data) == "event_mt"])
# window_me_lm <- lm(event_me ~ ., data=window_data[!grepl("event_", colnames(window_data)) | colnames(window_data) == "event_me"])
```


```{r}
library("BAS")

bas_cast_window_data <- as.data.frame(cast_window_data[!grepl("event_", colnames(cast_window_data)) | colnames(cast_window_data) == "event_hp_m2"])
# for(col in colnames(dt)) {
#   dt[[col]][is.na(dt[[col]])] <- mean(dt[[col]], na.rm = TRUE)
# }
# any(is.na(dt))
set.seed(12345)
cast_window_bas <- bas.lm(log1p(event_hp_m2) ~ ., data=bas_cast_window_data, n.models=2^14, method='BAS')
p <- bas_plot(cast_window_bas)

dt <- window_data[!grepl("event_", colnames(window_data)) | colnames(window_data) == "event_hp_m2"]
# for(col in colnames(dt)) {
#   dt[[col]][is.na(dt[[col]])] <- mean(dt[[col]], na.rm = TRUE)
# }
# any(is.na(dt))
set.seed(0)
window_bas <- bas.lm(event_hp_m2 ~ ., data=as.data.frame(dt), n.models=2^11, method='BAS')

mt_dt <- window_data[!grepl("event_", colnames(window_data)) | colnames(window_data) == "event_mt"]
mt_window_bas <- bas.lm(event_mt ~ ., data=as.data.frame(mt_dt), n.models=2^11, method='BAS')

me_dt <- window_data[!grepl("event_", colnames(window_data)) | colnames(window_data) == "event_me"]
me_window_bas <- bas.lm(event_me ~ ., data=as.data.frame(me_dt), n.models=2^11, method='BAS')
# me_window_bas <- bas.lm(log10(0.1 + event_me) ~ ., data=as.data.frame(me_dt), n.models=2^18, prior="ZS-null", modelprior=uniform(), method = "MCMC")
# image(me_window_bas, top.models = 50)# window_bas_subset <- bas.lm(event ~ ., data=window_data, n.models=2^20, prior='JZS')

# image(window_bas, top.models = 100)
# image(window_bas_subset)

# window_ela <- cv.glmnet(as.matrix(dt[, -79]), dt$event, family="gaussian")

# pca <- prcomp(as.matrix(window_data[, -79]), scale.=TRUE)
```

```{r}
bas_plot <- function (x, top.models = 20) {
  require(tidyverse)
  xcoef <- coef(x)
  incprob <- tibble(Predictor=xcoef$namesx, InclusionProb=xcoef$probne0)
  
  postprob <- x$postprobs
  top.models <- min(top.models, x$n.models)
  best <- order(-x$postprobs)[1:top.models]
  postprob <- postprob[best]/sum(postprob[best])
  which.mat <- list2matrix.which(x, best)
  nvar <- ncol(which.mat)
  subset <- 1:nvar
  which.mat <- which.mat[, subset, drop = FALSE]
  df <- as.data.frame(which.mat) %>%
    mutate(Rank=row_number()) %>%
    gather(Predictor, Included, 1:nvar) %>%
    left_join(tibble(Rank=1:top.models, PosteriorProb=postprob)) %>%
    mutate(PosteriorProb=case_when(Included == 0 ~ -1,
                                   TRUE ~ PosteriorProb)) %>%
    left_join(incprob) %>%
    group_by(Predictor) %>%
    mutate(PredIncludedMean=mean(Included == 1)) %>%
    ungroup() %>%
    arrange(desc(InclusionProb)) %>%
    mutate(Predictor=factor(Predictor, levels=rev(unique(.$Predictor))))
  p_left <- ggplot(df, aes(Rank, Predictor, fill=PosteriorProb)) +
    geom_tile() +
    scale_fill_gradient2(low="grey50", mid="white", high="red", guide=FALSE) +
    xlim(1, top.models) +
    theme_bw() +
    theme(panel.grid.major=element_blank(),
          panel.grid.minor=element_blank()) 
  p_right <- distinct(df, Predictor, InclusionProb) %>%
    ggplot(aes(Predictor, InclusionProb)) +
    geom_bar(stat='identity', fill='dodgerblue') +
    coord_flip() +
    theme_bw() +
    xlab("") +
    ylab("Inclusion Probability") +
    theme(panel.grid.major=element_blank(),
          panel.grid.minor=element_blank(),
          panel.border=element_blank(),
          axis.text.y=element_blank(),
          axis.line=element_line(size=0.2)
          )
    
  p <- gridExtra::grid.arrange(p_left, p_right, nrow=1, widths=c(80, 25))
  return(p)  
}

p <- bas_plot(cast_window_bas)
p_mt <- bas_plot(mt_window_bas)
p_me <- bas_plot(me_window_bas)
ggsave("~/Desktop/sci-lianti-files/bas_cast_varimp.pdf", p, "pdf", width=8, height=9)
ggsave("~/Desktop/sci-lianti-files/bas_varimp_mt.pdf", p_mt, "pdf", width=8, height=9)
ggsave("~/Desktop/sci-lianti-files/bas_varimp_me.pdf", p_me, "pdf", width=8, height=9)

sink("~/Desktop/sci-lianti-files/bas_cast_coef.txt")
print(coef(cast_window_bas))
sink()
```

# PCA

```{r}
cast_cell_pca <- prcomp((cast_cell_df_with_cor[-(1:2)]), scale.=T)
cast_cell_pca_noupd <- prcomp((cast_cell_df_with_cor[!colnames(cast_cell_df_with_cor) %in% upd_features][-(1:2)]), scale.=T)
cast_cell_pca_nomtme <- prcomp((cast_cell_df_with_cor[!colnames(cast_cell_df_with_cor) %in% mtme_features][-(1:2)]), scale.=T)
cast_cell_pca_noupdmtme <- prcomp((cast_cell_df_with_cor[!colnames(cast_cell_df_with_cor) %in% c(upd_features, mtme_features)][-(1:2)]), scale.=T)

png("~/Desktop/sci-lianti-files/cast_cell_pca.png", 8000, 6000, res=200)
do.call("grid.arrange", c(plot_cell_pca(cast_cell_pca, cast_cell_df_with_cor, "PC1", "PC3"), nrow=9))
dev.off()

cast_sub_cols <- c("TYPE", "chr3_bp","chr1_upd","QUANTILE_25_75_FLAG","CTCF","RNA_POL2","H3K27AC","H3K36ME3","DMRT6","H3K27ME3","PS_H3K27AC","PS_COHESIN","PS_SMC1B_LAP","CTCFL","END_SEQ_B6_ETO","RAD21_B6","PATSKI_ALLELIC_ATAC_B6","PATSKI_ALLELIC_CTCF_B6","CNV_CAST_LOSS","KAC")

png("~/Desktop/sci-lianti-files/cast_cell_pca_selected_cols.png", 6000, 4000, res=200)
do.call("grid.arrange", c(plot_cell_pca(cast_cell_pca, cast_cell_df_with_cor, "PC1", "PC3", cols=cast_sub_cols), ncol=5))
dev.off()


png("~/Desktop/fen_stuff/cast_cell_pca_noupd.png", 6000, 6000, res=200)
do.call("grid.arrange", c(plot_cell_pca(cast_cell_pca_noupd, cast_cell_df_with_cor), nrow=9))
dev.off()
png("~/Desktop/fen_stuff/cast_cell_pca_nomtme.png", 6000, 6000, res=200)
do.call("grid.arrange", c(plot_cell_pca(cast_cell_pca_nomtme, cast_cell_df_with_cor), nrow=9))
dev.off()
png("~/Desktop/fen_stuff/cast_cell_pca_noupdmtme.png", 6000, 6000, res=200)
do.call("grid.arrange", c(plot_cell_pca(cast_cell_pca_noupdmtme, cast_cell_df_with_cor), nrow=9))
dev.off()

ggsave("~/Desktop/sci-lianti-files/cast_cell_pca_pc1_pc3.pdf", plot_cell_pca_simp(cast_cell_pca, cast_cell_df_with_cor, "PC1", "PC3"), "pdf", width=4, height=3)
ggsave("~/Desktop/fen_stuff/cell_pca_noupd.pdf", plot_cell_pca_simp(cast_cell_pca_noupd, cast_cell_df_with_cor), "pdf", width=4, height=3)
ggsave("~/Desktop/fen_stuff/cell_pca_nomtme.pdf", plot_cell_pca_simp(cast_cell_pca_nomtme, cast_cell_df_with_cor), "pdf", width=4, height=3)
ggsave("~/Desktop/fen_stuff/cell_pca_noupdmtme.pdf", plot_cell_pca_simp(cast_cell_pca_noupdmtme, cast_cell_df_with_cor), "pdf", width=4, height=3)
```

# Correlation among window features

```{r}
# event_hp, event_m2, event_mt, event_me

window_cast_features_for_cor <- window_cast_with_cor %>%
  select(-c(CHROM, START, END, DISTANCE, BP_MID, CHROM_SIZE, QUANTILE_25, QUANTILE_50, QUANTILE_75)) %>%
  mutate_if(sapply(., is.character), function(x) {ifelse(x == "Yes", 1, 0)})

window_cast_features_cor <- cor(window_cast_features_for_cor, use='pairwise.complete.obs', method='spearman')
dim(window_cast_features_cor)
window_cast_features_order <- corrplot::corrMatOrder(window_cast_features_cor[1:68, 1:68], order = "hclust", hclust.method = "complete")
window_cast_features_order <- c(c(69:73), window_cast_features_order)

pdf("~/Desktop/sci-lianti-files/event_feature_window_cor_cast.pdf", 10, 10)
corrplot::corrplot(window_cast_features_cor[window_cast_features_order, window_cast_features_order], tl.cex=0.5, type="upper", method="color", tl.col=c(rep("red", 5), rep("steelblue", 78)))
dev.off()

window_features_cor[grep("event", colnames(window_features_cor), value=T), ] %>%
  as_tibble(rownames="EVENT") %>%
  write_tsv("~/Desktop/sci-lianti-files/event_feature_window_cor.tab")
```

# Apr 11 address these:

1. crossover pileup correlation matrix (Fig.S7)
need Pearson correlation r and p-value
event_haploid ~ event_m2
event_haploid+event_m2 ~ spo11_both
event_haploid+event_m2 ~ spo11_both + spo11_b6 + spo11_spret + spo11_other

```{r}
c1 <- cor.test(window_features_for_cor$event_hp, window_features_for_cor$event_m2, method="spearman")
c1$estimate
c1$p.value

c2 <- cor.test(window_features_for_cor$event_hp + window_features_for_cor$event_m2, window_features_for_cor$SPO11_BOTH, method="spearman")
c2$estimate
c2$p.value

c3 <- cor.test(window_features_for_cor$event_hp + window_features_for_cor$event_m2,
         window_features_for_cor$SPO11_BOTH + window_features_for_cor$SPO11_B6 + window_features_for_cor$SPO11_SPRET + window_features_for_cor$SPO11_OTHER, method="spearman")
c3$estimate
c3$p.value
```

2. crossover pileup regression:
lm summary table (coefficient and p-value)
summary table (MIP for each feature) from BMA (edited)

```{r}
summary(window_hp_m2_lm)
coef(window_bas)
```

3. “cell_pca” all feature plot, not only PC1 and PC2 with haploid and m2 diploid separtion
4. Random forest 0.73 AUC with all features, want an AUC with only features included > median inclusion probability from BMA, or only with features significant after multiple correction from lm.

done above.

4. Random forest 0.73 AUC with all features, want an AUC with only features included > median inclusion probability from BMA, or only with features significant after multiple correction from lm.

```{r}
bma_coef <- coef(window_bas)
bma_top_predictors <- setdiff(bma_coef$namesx[bma_coef$probne0 > median(bma_coef$probne0)], "Intercept")

lm_coef <- anova(window_hp_m2_lm)
lm_top_predictors <- setdiff(rownames(lm_coef)[lm_coef$`Pr(>F)` < 0.05], c("Intercept", NA))

sub_md_data <- md_data[c("RESPONSE", "TYPE", bma_top_predictors)]

# ncv_list <- bplapply(1:1, function(x) {
  train_n <- round(nrow(sub_md_data) * 0.9)
  train_idx <- sample(1:nrow(sub_md_data), size=train_n, replace=FALSE)
  train_df <- sub_md_data[train_idx, ]
  train_df$RESPONSE <- factor(train_df$RESPONSE)
  test_df <- sub_md_data[-train_idx, ]
  train_control <- trainControl(method = "cv", number = 5, returnResamp = "all",
                                classProbs = TRUE, 
                                summaryFunction = twoClassSummary)
  model_weights <- ifelse(train_df$RESPONSE == "Y",
                          (1/table(train_df$RESPONSE)["Y"]) * 0.5,
                          (1/table(train_df$RESPONSE)["N"]) * 0.5)
  md <- train(RESPONSE ~ ., data = train_df, 
              method = "rf", 
              trControl = train_control,
              weights = model_weights,
              metric = "ROC")
  pred <- predict(md, newdata=test_df, type = "prob")
  pred_df <- tibble(truth=test_df$RESPONSE, pred=pred$Y)
#  return(pred_df)
# })

sub_md_data2 <- md_data[c("RESPONSE", "TYPE", lm_top_predictors)]

# ncv_list <- bplapply(1:1, function(x) {
  train_n <- round(nrow(sub_md_data2) * 0.9)
  train_idx <- sample(1:nrow(sub_md_data2), size=train_n, replace=FALSE)
  train_df <- sub_md_data2[train_idx, ]
  train_df$RESPONSE <- factor(train_df$RESPONSE)
  test_df <- sub_md_data2[-train_idx, ]
  train_control <- trainControl(method = "cv", number = 5, returnResamp = "all",
                                classProbs = TRUE, 
                                summaryFunction = twoClassSummary)
  model_weights <- ifelse(train_df$RESPONSE == "Y",
                          (1/table(train_df$RESPONSE)["Y"]) * 0.5,
                          (1/table(train_df$RESPONSE)["N"]) * 0.5)
  md2 <- train(RESPONSE ~ ., data = train_df, 
              method = "rf", 
              trControl = train_control,
              weights = model_weights,
              metric = "ROC")
  pred2 <- predict(md2, newdata=test_df, type = "prob")
  pred_df2 <- tibble(truth=test_df$RESPONSE, pred=pred$Y)  
 
roc_obj <- roc(pred_df2$truth, pred_df2$pred)
plot(roc_obj)
auc(roc_obj)
```
